src/climate_learn/data/climate_dataset/era5_module.py

```diff
@@ -1,94 +1,235 @@
-import os
+# Standard library
+import copy
 import glob
-import xarray as xr
-import numpy as np
+import os
+import random
+from typing import Callable, Dict, Iterable, Sequence, Tuple
 
+# Third party
+import numpy
+import torch
 from tqdm import tqdm
-from typing import Callable, Dict, Iterable, Sequence
-from climate_learn.data.climate_dataset import ClimateDataset
+import xarray as xr
+
+# Local application
 from climate_learn.data.climate_dataset.args import ERA5Args
-from ..constants import (
-    NAME_TO_VAR,
-    DEFAULT_PRESSURE_LEVELS,
+from climate_learn.data.climate_dataset import ClimateDataset
+from climate_learn.data.constants import (
     CONSTANTS,
-    SINGLE_LEVEL_VARS,
+    DEFAULT_PRESSURE_LEVELS,
+    NAME_LEVEL_TO_VAR_LEVEL,
+    NAME_TO_VAR,
     PRESSURE_LEVEL_VARS,
+    SINGLE_LEVEL_VARS,
 )
 
 
 class ERA5(ClimateDataset):
-    args_class: Callable[..., ERA5Args] = ERA5Args
+    _args_class: Callable[..., ERA5Args] = ERA5Args
 
     def __init__(self, data_args: ERA5Args) -> None:
         super().__init__(data_args)
         self.root_dir: str = data_args.root_dir
         self.years: Iterable[int] = data_args.years
 
-    def setup(self) -> None:
-        self.constant_names: Sequence[str] = []
-        self.data_dict: Dict[str, xr.core.dataarray.DataArray] = self.load_from_nc(
-            self.root_dir
+    def get_file_name_from_variable(self, var: str) -> str:
+        if var in SINGLE_LEVEL_VARS or var in PRESSURE_LEVEL_VARS:
+            return var
+        else:
+            return "_".join(var.split("_")[:-1])
+
+    def set_lat_lon(self, year: int) -> None:
+        # lat lon is stored in each of the nc files, just need to load one and extract
+        dir_var = os.path.join(
+            self.root_dir, self.get_file_name_from_variable(self.variables[0])
         )
-        self.set_lat_lon()
+        ps = glob.glob(os.path.join(dir_var, f"*{year}*.nc"))
+        xr_data = xr.open_mfdataset(ps, combine="by_coords")
+        self.lat: numpy.ndarray = xr_data["lat"].values
+        self.lon: numpy.ndarray = xr_data["lon"].values
 
-    def load_from_nc(self, data_dir: str) -> Dict[str, xr.core.dataarray.DataArray]:
-        self.constant_names = [
-            name for name in self.variables if NAME_TO_VAR[name] in CONSTANTS
+    def setup_metadata(self, year: int) -> None:
+        ## Prevent setup if already called before
+        if hasattr(self, "lat") and hasattr(self, "lon"):
```

Comment:
The `ShardDataset` calls` setup()` a lot in fact every worker of `DataLoader` in every epoch calls it. Since, `metadata` doesn't really change irrespective of chunk in the case of `ERA5`, why waste time in opening in a file. I/O can be a bottleneck especially if we are repeatedly reading files.

Thus I wanted to `setup_metadata()` just once.

Reply:
Sorry, I wasn't clear in my comment. It's fine that `lat` and `lon` are set in the `setup_metadata` method. I'm saying that I prefer that they are declared and set in two different places: declaration in `__init__` and set in `setup_metadata`. Thus, if we already declared the variables in `__init__`, instead of checking `hasattr`, we can just check if the variable is set to something other than `None`.

Reply:
Sure. We can do what you suggested.

Reply:
Addressed.